Fix make disk image automatic size

samueldr opened the pull request:
###### Motivation for this change

 - #89331
 - #107212
 - #120481

This is doing the actual accounting we need to do to make the partition big enough to fit data.

This was tested by setting `additionalSpace` to `0`. Any other values (like the default 512MiB) can hide subtle issues. But the default 512MiB is fine, since it also hardens this implementation from small deviations.

The accounting fixes here comes from work I did for Mobile NixOS.

 - https://github.com/NixOS/mobile-nixos/tree/master/lib/image-builder

See also:

 - https://github.com/NixOS/nixpkgs/blob/93de027ee045d0c1fbf13418fb86d86e6309b962/nixos/lib/make-ext4-fs.nix#L58-L63

Don't forget to read the commit messages!

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Lassulus committed:
treewide: use auto diskSize for make-disk-image

(cherry picked from commit f3aa040bcbf39935e7e9ac7a7296eac9da7623ec)

samueldr committed:
make-disk-image: Get proper size for automatic size

On some filesystems, `du` without `--apparent-size` will not give the
actual size for a file. Using `--apparent-size` will give us the actual
file size.

Though, this is not actually correct still. 1000 × 1 bytes is not 1000
bytes. It is 1000 × ceil(filesize/blockSize)*blockSize.

So instead of adding up the actual file sizes. We are adding up the
block sizes.

Note that this also changes the builder to work with *bytes*, rather
than with any other units. Doing maths on bytes is less likely to go
awry than doing it on other units.

samueldr committed:
make-disk-image: Account for the ext4 reserved space

Reserved space includes:

 - inodes space in use (2 blocks per)
 - about 5.2% of the space

The 5.2% reserved space was computed empirically when working on a
previous EXT4 image builder. It seems to stabilize around 5% even for
much larger filesystems.

samueldr committed:
make-disk-image: Account for reserved disk space

This is a bit of a thorny issue. See, the actual `diskSize` variable is
for the *total* disk size, not for the filesystem!

The automatic numbers are meant to compute the *filesystem* required
space. So we have to add any other reserved space!

We have different requirements for reserved space. E.g. there could be
none (when it's actually a filesystem image). There could also be 1MiB
for alignment for an MBR image, legacy+gpt needs 2MiB, then GPT with an
ESP ("bootSize") needs to take the boot partition and GPT size into
account too!

Though luckily(?) for this latter situation we can cheat! As noted in the
change, `bootSize` is NOT the boot partition size. It is actually the
offset where the target filesystem starts.

Mic92 commented:
I guess this could also revert the revert if it works? https://github.com/NixOS/nixpkgs/pull/120487
I could test this on zfs with compression enabled.

lukegb commented:
Yeah, this should probably revert the revert too - or, probably better, re-cherry-pick f3aa040bcbf39935e7e9ac7a7296eac9da7623ec?

Lassulus reviewed (approved):
code looks very clean and is good documented, thanks a lot for fixing my mess :)

I would be very happy to have the dynamic sized images in 21.05 if everything works

samueldr commented:
Since it could get confusing "how did you test this if the commits were already reverted?"

Well, I was based off a commit from before the reverts merge PR :).

Mic92 commented:
I tried to build openstack with nixos-generators, but it segfaulted my nix. This is probably unrelated to changes here.

numinit commented:
@Mic92 I ran into the same thing. I just happened to be concurrently refactoring make-disk-image.nix to allow for creating swap space after the boot partition and rebased against this PR. Tried `nix-shell -p nix` to get the non-experimental nix and everything worked afterwards.

Anyway, this change looks good to me after spending a Saturday afternoon familiarizing myself with this part of nixpkgs. :+1:

[The swap space changes are in preparation for trying to add ZFS support to it too - my idea is automatically provisioning NixOS images that look [like this](https://grahamc.com/blog/nixos-on-zfs). I have lots of servers where the dataset layout @grahamc  describes works _really_ well, and would love to just build an image that I can dd straight to the disk instead of running a long playbook of commands over and over again for every server with minor differences for every one...]

samueldr commented:
@numinit WAIT! collaborate on #106574 !

numinit commented:
@samueldr Nice. Didn't know about that one. I've not done any work on the ZFS stuff yet, but will still PR the swapspace changes because they're useful anyway here. :-)

numinit commented on nixos/lib/make-disk-image.nix:
```diff
@@ -235,12 +255,45 @@ let format' = format; in let
+      fudge=$(compute_fudge $diskUsage)
+      requiredFilesystemSpace=$(( diskUsage + fudge ))
+
+      diskSize=$(( requiredFilesystemSpace  + additionalSpace ))
```
This isn't necessarily a multiple of 512 bytes and makes fdisk/gdisk complain. parted seems to ignore it, but probably still a good idea to round up to the next highest blockSize here.

samueldr replied:
This probably would be wrong.

What I assume it complains about is the partitions starting/ending unaligned, and not necessarily on sector size (commonly 512 bytes). I believe the alignment by default is 1MiB? Though alignment is actually a property of the block device being partitioned! Doing any alignment on an image might actually end up wrong!

For this, we'd need to re-do the whole disk image build logic to actually work with discrete filesystem images that we *then* place as partitions into a larger disk image.

Right now, we're actually trying to carve a hole into a bigger image we *hope* we're making big enough. Well, we're doing less of it with this PR.

Though I guess rounding up to the nearest megabyte is a clean way to avoid this warning.

numinit replied:
It's actually complaining about file size (three whole times for some reason). And it's just gdisk.

```
$ truncate -s 1234 test
$ gdisk test
GPT fdisk (gdisk) version 1.0.5

Warning: File size is not a multiple of 512 bytes! Misbehavior is likely!
Warning: File size is not a multiple of 512 bytes! Misbehavior is likely!
Warning: File size is not a multiple of 512 bytes! Misbehavior is likely!
```

Yeah, rounding up to the nearest MB would fix it. I think parted actually recommends that everything be 4MiB aligned at least. Best we can do is guess here anyway and 1 MB alignment seems good enough for most cases.

samueldr replied:
Right, this warning is probably a non-issue if it is on a disk image:

 - https://sourceforge.net/p/gptfdisk/code/ci/eae9f7e2e1470ed2d5bc029180aa2adc88fbc4a7/tree/diskio-unix.cc#l496
 - https://sourceforge.net/p/gptfdisk/code/ci/eae9f7e2e1470ed2d5bc029180aa2adc88fbc4a7/tree/diskio-unix.cc#l493

Hmm... Though it would mean that the backup GPT at the end of the disk is not at LBAEND-33, but rather LBAEND-34, since it rounded down the amount of sectors. BUT, again this is a non-issue since when `dd`'ing the disk image to the actual device, the device will always be bigger, and that the backup GPT will invariably not be at the end sector.

samueldr replied:
And it is done

https://github.com/NixOS/nixpkgs/pull/120489/commits/7b8b3fab6dce5357c14fc435b436514fc985f0f7

samueldr committed:
make-disk-image: Round image size to the next mebibyte

This ensures the following gptfdisk warning won't happen:

```
Warning: File size is not a multiple of 512 bytes! Misbehavior is likely!
```

Additionally, helps towards aligning the partition to be more optimal
for the underlying storage.

It is actually impossible to align for the actual underlying storage
optimally because we don't know what the block device will be!

But aligning on 1MiB should help.

numinit commented on nixos/lib/make-disk-image.nix:
```diff
@@ -235,12 +255,45 @@ let format' = format; in let
+      additionalSpace=$(( $(numfmt --from=iec '${additionalSpace}') + reservedSpace ))
+
+      # Compute required space in filesystem blocks
+      diskUsage=$(find . ! -type d -exec 'du' '--apparent-size' '--block-size' "${blockSize}" '{}' ';' | cut -f1 | sum_lines)
```
Do we need to add size for directories (plus their entries) and symlinks, or do you think this is a "good enough" estimate, those aside?

samueldr replied:
This inodes accounting comes from a contributor who's not contributing anymore. I don't know the details. But I assume they take as much space as files. Directories and symlinks have distinct inodes when listing them with `ls -i`.

vcunat commented:
Some of you might know... [one image job](https://hydra.nixos.org/build/142244712) now started blocking `nixos-unstable` channel due to producing a too large output (slightly over the 2 GiB limit).

samueldr commented:
Last successful:

```
irb(main):008:0> 3128950784.0 / 1024 / 1024 / 1024
=> 2.9140625
irb(main):009:0> 3128950784.0 / 1000 / 1000 / 1000
=> 3.128950784
```

```
irb(main):002:0> 3128950784.0 / 1024 / 1024 / 1024
=> 2.9140625
irb(main):003:0> 3128950784.0 / 1000 / 1000 / 1000
=> 3.128950784
```

So uh... Odd!

The **reported** disk image size is the same from the changes from this PR.

It might be that the actual file being produced is different.

samueldr commented:
```
[20:25:29] <samueldr> max output size is 2147483648
[20:27:16] <samueldr> is that once the NAR is compressed?
[20:27:59] <lukegb> yes
[20:31:26] <samueldr> right
[20:31:48] <samueldr> still, odd that it's over only sometimes?
[20:32:34] <lukegb> I guess building the image is non-deterministic
[20:32:41] <lukegb> and we're _just_ close to the limit
[20:34:58] <samueldr> ~1705414492.16 bytes
[20:35:28] <samueldr> I would be surprised that it is *that much* irreproducible
[20:35:42] <samueldr> output size from https://hydra.nixos.org/build/142182234#tabs-details
```

Lassulus commented:
hmm, I guess we can set the filesize of the amazon image to unblock the channel? maybe we can set it to auto again when we reduce the size of the image somehow of raise the limit?

vcunat commented:
Unblocked by PR #121352.
